runtime/lua/vim/shared.lua

```diff
@@ -1139,4 +1139,24 @@ function vim._defer_require(root, mod)
   })
 end
 
+--- @private
+function vim._with(context, f)
+  -- vim._with({ options = { isfname = vim.o.isfname .. ',@-@' } }, function() end)
+  local previous_options = {}
+  if context.options then
+    for k, v in pairs(context.options) do
+      previous_options[k] = vim.o[k]
+      vim.o[k] = v
```

Comment:
```suggestion
      vim.api.nvim_set_option_value(k, v, { scope = 'local', win = context.win, buf = context.buf })
```

I think this should take care of setting the properly scoped option values whenever `win/buf/neither` is passed? Same with `nvim_get_option_value`.

Reply:
This is exactly the sort of thing we need to test for. Changing via `vim.o` won't work properly if `buf` or `win` is passed.

Reply:
Unsure what you mean, running

```
vim._with({buf = 1, options = { spelllang = "fr" }}, function() end)
vim.cmd[[set spelllang?]]
```

prints `en` which is the expected value.

Reply:
Yes, but now try it with a `win` context.

Reply:
OK, I could recreate the problem. We can't use `scope` together with `buf` though so will need to throw that out.